### ISSUE TITLE
Allow creating custom resources when creating namespaces in profile controller

### ISSUE THREAD

**Issue by discordianfish** [kubeflow/kubeflow#4215]
/kind feature

**Why you need this feature:**
With the multi-user support, every user gets a new, blank namespace. To use secrets or access podpresets in the jupyter UI, an admin needs to apply manifests after the namespace creation. Since the namespace creation happens dynamically on first login, it's not possible to automate that. At least without setting up all profiles in advance which we don't want to do since it would require use to somehow synchronize kubeflow uses with our user directory.

**Describe the solution you'd like:**
The profile-controller already creates some resources on namespace creation: https://github.com/kubeflow/kubeflow/blob/a4286d524a18e86bda97b5fcf05fa4086063513a/components/profile-controller/pkg/controller/profile/profile_controller.go#L226
This should be extended to apply custom resources. To configure which resources to create, a few options come to mind:

- Add configmap or CRD with manifests to apply to each new namespace OR
- Copy all resources with well defined label from kubeflow namespace
  - Pro: Straight forward
  - Cons: Static, e.g can't support different manifests based on user

- Add webhook to call external service with Profile and retrieves resources to apply
  - Pro: Flexible
  - Cons: Non-trivial to setup and build in a robust way

- "Support" custom admission controller for this use case
  - Pro: Flexible, no changes to profile-controller
  - Cons: Very involved to build, but maybe a 'framework' could help

All these options could also support/preset namespace naming schema to skip UI step of configuring the workspace without having to create all namespaces in advance.

**Anything else you would like to add:**
- #2192 Is related, though specific to secret and was closed due to staleness.

**Comment by jlewi** [comment]
Cc @kunming

**Comment by discordianfish** [comment]
@jbottum Are there already ideas how to implement this? Do you want/need input on this or help with implementing? From my perspective it would be ideal if we can get around this initial workflow setup completely which requires similar configuration of the profile controller. Basically I want to configure:

- What namespace to create for a new user based on it's auth detail (e.g their email address)
- What resources to create, again ideally based on the auth detail (e.g I might want to label some resources with the users email to refer to in selectors and general for ops purposes)

Since they are closely related, it might make sense to address that as one.

**Comment by jlewi** [comment]
I have a couple of questions on how this would work

* Shouldn't secrets be different for each namespace? If every namespace gets provisioned with the same secrets then everyone has access to to the same sets of resources so doesn't that defeat the isolation that namespaces are providing?

* Why would an admin need to provision the secrets? Wouldn't a user have access to the secret and be able to provision it themselves?

**Comment by discordianfish** [comment]
> Shouldn't secrets be different for each namespace? If every namespace gets provisioned with the same secrets then everyone has access to to the same sets of resources so doesn't that defeat the isolation that namespaces are providing?

Ideally secrets shouldn't be shared but in reality you sometimes don't get around this. But even in an ideal world, I'd need to inject the secrets *somehow*, even if I generate new secrets for each user that sets up their workspace. But for the sake of argument, let's ignore secrets and rather talk about e.g podpresets. But I don't think kubeflow should have an opinion on what a admin wants to deploy there.

> Why would an admin need to provision the secrets? Wouldn't a user have access to the secret and be able to provision it themselves?

Ideally a user shouldn't need to setup anything themselves. A new data scientist should be able to login and start creating/running workflows that might require secrets to connect to 3rd party services or rely on podpresets to configure their notebooks.

Beside that, @jlewi what would it take for me to drive this? Is there a template for such design proposals? I'd like to get more involved in kubeflow development if that's an option at this point.

**Comment by jlewi** [comment]
@discordianfish My suggestion would be to write up a doc; we don't have a good template but the key questions would be

1. What is the use case / CUJ you are trying to solve
1. How you will implement it
1. Alternatives considered 

I would suggest sharing the doc with the mailing list and then adding it to the agenda of a community meeting to discuss.

**Comment by klolos** [comment]
Hello @discordianfish,

Are you currently working on this issue?

I would also like to start working on it, so if someone has made some progress please let me know so we can exchange ideas.

**Comment by discordianfish** [comment]
I thought about this some more and came up with: https://docs.google.com/document/d/1DvglZubJEvK3fuo3mi1bCpdW4hxK4zFkmVy0qTcepiI/edit?folder=12XkDmfUE8vCXP9-hp3ft9KadJjRxY_lZ# which I also shared on the mailinglist.

**Comment by jlewi** [comment]
@discordianfish I would suggest presenting this at one of the community meetings.

**Comment by discordianfish** [comment]
@jlewi I would but that's 2:30am for me (Berlin, Germany). Any way to do this async?

**Comment by yanniszark** [comment]
@discordianfish thanks a lot for taking the time to write down a proposal.

We (Arrikto) have also encountered this issue and we came up with a slightly different solution, which we believe is simpler and more effective.

In our solution, extra resources are defined in a ConfigMap which is given to the Profile Controller.
Each namespace belonging to a Profile is reconciled to make sure all the necessary resources are created.

The main advantages of this solution compared to NamespaceDefaults is that:
* Resources are reconciled constantly and not just on namespace creation (level-based vs edge-based).
* Implementation is straightforward, while with the NamespaceDefaults proposal there are details that need to be ironed out with no obvious solution. eg the namespace doesn't exist when the webhook gets called, so there needs to be some kind of scheduling mechanism.

We detail our proposal in the following doc: https://docs.google.com/document/d/1CQasIRkyqK-UEyi2COgg8bfyfydq7sbh17xjVNVLjcQ/edit?usp=sharing
We can also share our code so far, we have been using this for some of our deployments.

/cc @jlewi

**Comment by discordianfish** [comment]
@yanniszark Looks like this is close to what I proposed initially. But now thinking a CRD might be a more suitable solution, given how it aligns with PodPresets and the fact that it's typed and can be checked by an admission controller.

Anyway, I don't care much about the specific solution but it's frustrating to waste time on competing proposals. A lot of decisions and direction apparently is hard to follow. I know about the community meetings but again it's 2:30am for Europe. On the mailinglist and github I never get a response and on Slack only if I nag people beyond what I feel comfortable doing.

**Comment by jlewi** [comment]
@yanniszark @discordianfish what is the current status for this? I think there was some discussion of the Arrikto proposal at last week's meeting.

**Comment by discordianfish** [comment]
Wasn't at last week's meeting due to it's time.
But I actually like the idea of starting with ClusterPodPreset and figure out if we need a generic approach for other resources. In our case we'd still need a way to manage secrets, but I agree that we'd be better of to not having to share secrets in the first place, so would be reasonable to me to find another solution for that.

**Comment by jlewi** [comment]
@discordianfish maybe open up a separate issue to discuss the secret management problem and possible solutions?

Its a big problem for Kubeflow development as well as Kubeflow users. Lots of KF infrastructure depends on provisioning secrets and we don't have a good way of managing them. Right now we basically require users to manually create secrets with `kubectl` which doesn't really scale and limits our ability to fully automate things.

**Comment by discordianfish** [comment]
@jlewi @yanniszark What is the current state of this? Should I replace this issue by two:

1. Add ClusterPodPresets
2. Find solution for secret management

**Comment by jlewi** [comment]
@discordianfish see the discussion in:
https://groups.google.com/g/kubeflow-discuss/c/whS_j4RlEJo

And related issue kubeflow/manifests#1136 

I think the direction we are moving towards (at least for GCP) is to do namespace setup declaratively. Then its easy for platform teams to customize it for their specific needs by adding resources to packages that they duplicate.

Does that work for you?

**Comment by discordianfish** [comment]
@jlewi Hrmm.. it depends on the specifics, but let's discuss that in kubeflow/manifests#1136 and close this for now.

**Comment by abdulsalama** [comment]
> We can also share our code so far, we have been using this for some of our deployments.

@yanniszark do you mind sharing the code you have so far for this?
